File: backend/apps/navigation/schemas.py
from typing import Literal
from uuid import UUID

from ninja import Schema
from pydantic import Field


NavigationLocationValue = Literal[
    "header_primary",
    "header_secondary",
    "footer_primary",
    "footer_secondary",
    "footer_legal",
    "mobile",
    "dashboard",
    "client_portal",
    "custom",
]

MenuItemLinkTypeValue = Literal[
    "internal",
    "external",
    "route",
    "cms_page",
    "service",
    "package",
    "industry",
    "insight",
    "case_study",
    "careers",
    "contact",
    "quote",
    "custom",
]

MenuItemVisibilityValue = Literal[
    "everyone",
    "guests",
    "authenticated",
    "staff",
    "superuser",
]


class NavigationMenuCreateSchema(Schema):
    name: str
    slug: str
    location: NavigationLocationValue = "custom"
    description: str = ""
    is_active: bool = True
    is_public: bool = True
    sort_order: int = 0
    metadata: dict = Field(default_factory=dict)


class NavigationMenuUpdateSchema(
    NavigationMenuCreateSchema
):
    pass


class NavigationMenuItemCreateSchema(Schema):
    parent_id: UUID | None = None
    label: str
    link_type: MenuItemLinkTypeValue = "internal"
    url: str = ""
    route_name: str = ""
    route_parameters: dict = Field(
        default_factory=dict
    )
    cms_page_id: UUID | None = None
    service_id: UUID | None = None
    package_id: UUID | None = None
    industry_id: UUID | None = None
    insight_id: UUID | None = None
    case_study_id: UUID | None = None
    visibility: MenuItemVisibilityValue = "everyone"
    icon: str = ""
    css_class: str = ""
    target_blank: bool = False
    rel_attribute: str = ""
    is_active: bool = True
    is_featured: bool = False
    sort_order: int = 0
    metadata: dict = Field(default_factory=dict)


class NavigationMenuItemUpdateSchema(
    NavigationMenuItemCreateSchema
):
    pass


class NavigationReorderItemSchema(Schema):
    id: UUID
    parent_id: UUID | None = None
    sort_order: int


class NavigationReorderSchema(Schema):
    items: list[
        NavigationReorderItemSchema
    ] = Field(default_factory=list)


class NavigationMenuItemSchema(Schema):
    id: UUID
    menu_id: UUID
    parent_id: UUID | None
    label: str
    link_type: str
    url: str
    resolved_url: str
    route_name: str
    route_parameters: dict
    cms_page_id: UUID | None
    service_id: UUID | None
    package_id: UUID | None
    industry_id: UUID | None
    insight_id: UUID | None
    case_study_id: UUID | None
    visibility: str
    icon: str
    css_class: str
    target_blank: bool
    rel_attribute: str
    is_active: bool
    is_featured: bool
    sort_order: int
    depth: int
    metadata: dict
    children: list["NavigationMenuItemSchema"]


class NavigationMenuSchema(Schema):
    id: UUID
    name: str
    slug: str
    location: str
    description: str
    is_active: bool
    is_public: bool
    sort_order: int
    metadata: dict
    item_count: int
    items: list[NavigationMenuItemSchema]



class PublicNavigationMenuItemSchema(Schema):
    id: UUID
    parent_id: UUID | None
    label: str
    link_type: str
    url: str
    icon: str
    css_class: str
    target_blank: bool
    rel_attribute: str
    is_featured: bool
    sort_order: int
    metadata: dict
    children: list[
        "PublicNavigationMenuItemSchema"
    ]


class PublicNavigationMenuSchema(Schema):
    id: UUID
    name: str
    slug: str
    location: str
    description: str
    metadata: dict
    items: list[PublicNavigationMenuItemSchema]

File: apps/dashboard/src/features/navigation/schemas.ts
import {
  z,
} from "zod";

import {
  navigationLinkTypes,
  navigationLocations,
  navigationVisibilityValues,
  type NavigationMenuItem,
  type PublicNavigationMenuItem,
} from "./types";

const metadataSchema = z.record(
  z.string(),
  z.unknown(),
);

export const navigationMenuItemSchema:
z.ZodType<NavigationMenuItem> = z.lazy(
  () =>
    z.object({
      id: z.string().uuid(),
      menu_id: z.string().uuid(),
      parent_id:
        z.string().uuid().nullable(),
      label: z.string(),
      link_type: z.enum(
        navigationLinkTypes,
      ),
      url: z.string(),
      resolved_url: z.string(),
      route_name: z.string(),
      route_parameters: metadataSchema,
      cms_page_id:
        z.string().uuid().nullable(),
      service_id:
        z.string().uuid().nullable(),
      package_id:
        z.string().uuid().nullable(),
      industry_id:
        z.string().uuid().nullable(),
      insight_id:
        z.string().uuid().nullable(),
      case_study_id:
        z.string().uuid().nullable(),
      visibility: z.enum(
        navigationVisibilityValues,
      ),
      icon: z.string(),
      css_class: z.string(),
      target_blank: z.boolean(),
      rel_attribute: z.string(),
      is_active: z.boolean(),
      is_featured: z.boolean(),
      sort_order: z.number().int(),
      depth: z.number().int(),
      metadata: metadataSchema,
      children: z.array(
        navigationMenuItemSchema,
      ),
    }),
);

export const navigationMenuSchema =
  z.object({
    id: z.string().uuid(),
    name: z.string(),
    slug: z.string(),
    location: z.enum(
      navigationLocations,
    ),
    description: z.string(),
    is_active: z.boolean(),
    is_public: z.boolean(),
    sort_order: z.number().int(),
    metadata: metadataSchema,
    item_count: z.number().int(),
    items: z.array(
      navigationMenuItemSchema,
    ),
  });

export const publicNavigationItemSchema:
z.ZodType<PublicNavigationMenuItem> =
  z.lazy(() =>
    z.object({
      id: z.string().uuid(),
      parent_id:
        z.string().uuid().nullable(),
      label: z.string(),
      link_type: z.enum(
        navigationLinkTypes,
      ),
      url: z.string(),
      icon: z.string(),
      css_class: z.string(),
      target_blank: z.boolean(),
      rel_attribute: z.string(),
      is_featured: z.boolean(),
      sort_order: z.number().int(),
      metadata: metadataSchema,
      children: z.array(
        publicNavigationItemSchema,
      ),
    }),
  );

export const publicNavigationMenuSchema =
  z.object({
    id: z.string().uuid(),
    name: z.string(),
    slug: z.string(),
    location: z.enum(
      navigationLocations,
    ),
    description: z.string(),
    metadata: metadataSchema,
    items: z.array(
      publicNavigationItemSchema,
    ),
  });

export const navigationMenuPayloadSchema =
  z.object({
    name: z.string().trim().min(
      1,
      "Menu name is required.",
    ),
    slug: z.string().trim().min(
      1,
      "Menu slug is required.",
    ),
    location: z.enum(
      navigationLocations,
    ),
    description: z.string(),
    is_active: z.boolean(),
    is_public: z.boolean(),
    sort_order: z.number().int(),
    metadata: metadataSchema,
  });

export const navigationItemPayloadSchema =
  z.object({
    parent_id:
      z.string().uuid().nullable(),
    label: z.string().trim().min(
      1,
      "Item label is required.",
    ),
    link_type: z.enum(
      navigationLinkTypes,
    ),
    url: z.string(),
    route_name: z.string(),
    route_parameters: metadataSchema,
    cms_page_id:
      z.string().uuid().nullable(),
    service_id:
      z.string().uuid().nullable(),
    package_id:
      z.string().uuid().nullable(),
    industry_id:
      z.string().uuid().nullable(),
    insight_id:
      z.string().uuid().nullable(),
    case_study_id:
      z.string().uuid().nullable(),
    visibility: z.enum(
      navigationVisibilityValues,
    ),
    icon: z.string(),
    css_class: z.string(),
    target_blank: z.boolean(),
    rel_attribute: z.string(),
    is_active: z.boolean(),
    is_featured: z.boolean(),
    sort_order: z.number().int(),
    metadata: metadataSchema,
  });
